api: add add_external_repo and rm_external_repo

ryanfortner opened the pull request:
As suggested in https://github.com/Botspot/pi-apps/pull/1889, this is a function that adds external apt repositories when provided:

- an external repo path (such as `https://ryanfortner.github.io/box64-debs/debian ./`) 
- a repo name (such as `box64`)
- a url to the repo's gpg pubkey (such as `https://ryanfortner.github.io/box64-debs/KEY.gpg`)

Example of the usage of these functions:
```bash
add_external_repo "https://ryanfortner.github.io/box64-debs/debian ./" "box64" "https://ryanfortner.github.io/box64-debs/KEY.gpg" || exit 1

rm_external_repo "box64" || exit 1 # this will remove the repo if nothing is used in it
rm_external_repo "box64" "force" || exit 1 # this will remove the repo even if software is installed from it
```

This is my first try at creating a function for pi-apps and I would appreciate any feedback I receive!

theofficialgman commented:
We already have a remove repo if unused function. You should extend that to remove the key if used as well.
https://github.com/Botspot/pi-apps/blob/d20a6330924ce6523612717e7aec33b965a807fc/api#L214

Botspot commented on api:
```diff
@@ -351,6 +351,38 @@ app_to_pkgname() { #given an app-name, convert it to a unique, valid package-nam
 }
 
+add_external_repo() { # add an external apt repo and its gpg key
+  repopath="$1"
+  reponame="$2"
+  pubkeyurl="$3"
```
I recommend adding something to check that each of these values are populated, and that the URLs are reachable.

ryanfortner replied:
I have added a check for each of the variables, as well as a check for if `pubkeyurl` is reachable.

However I'm not sure how we could check if `repopath` is reachable... maybe test if the `Packages` file exists on the server?

Botspot commented on api:
```diff
@@ -351,6 +351,38 @@ app_to_pkgname() { #given an app-name, convert it to a unique, valid package-nam
+  [ -f /etc/apt/sources.list.d/${reponame}.list ] && error "add_external_repo: /etc/apt/sources.list.d/${reponame}.list already exists."
+
+  # download gpg key from specified url
+  rm -f /tmp/pi-apps-apt-keys/${reponame}-keyring.gpg
+  wget -q "$pubkeyurl" -O /tmp/pi-apps-apt-keys/${reponame}-keyring.gpg || error "add_external_repo: download from specified pubkeyurl failed." && rm -f /tmp/pi-apps-apt-keys/${reponame}-keyring.gpg
```
I recommend testing the conditionals in this line.
If wget succeeds, the `||` will be skipped but the `&&` will run `rm`, which immediately removes the file `wget` just downloaded. Is that intended?
Conversely, if wget fails, the `||` will trigger, which runs the `error` function and exit the script. In this case, the `rm` command won't run, when I think you intended it to run.

ryanfortner replied:
My intention was that if wget fails, then the incomplete download file should be removed as well as the error printed.

I have removed the && part anyways, because the function already checks if the url is reachable so it isn't likely that there will be download issues.

Botspot commented on api:
```diff
@@ -351,6 +351,38 @@ app_to_pkgname() { #given an app-name, convert it to a unique, valid package-nam
+  wget -q "$pubkeyurl" -O /tmp/pi-apps-apt-keys/${reponame}-keyring.gpg || error "add_external_repo: download from specified pubkeyurl failed." && rm -f /tmp/pi-apps-apt-keys/${reponame}-keyring.gpg
+
+  # create .list file
+  echo "deb [signed-by=/tmp/pi-apps-apt-keys/$reponame-keyring.gpg] $repopath" | sudo tee /etc/apt/sources.list.d/${reponame}.list >/dev/null || error "add_external_repo: failed to create .list file"
```
You seem to be treating files in /tmp as though they are persistent. They are not - every once in a while old contents will be cleared, and a reboot will clear it too.
GPG keys should be stored in a persistent, root-owned location.

ryanfortner replied:
I changed the location from `/tmp` to `/opt/pi-apps-apt-keys`.

Botspot replied:
@theofficialgman, where are system-wide GPG keys stored for things like apt repositories? Wouldn't it be better to store these keys in a standard location than a new one?

Crilum replied:
> @theofficialgman, where are system-wide GPG keys usually stored for things like apt repositories? Wouldn't it be better to store these keys in a standard location than a new one?

It looks like `/usr/share/keyrings/`.. Don't quote me on that.. 😆

Jai-JAP replied:
Yup thats the correct location for all apt keyrings

ryanfortner replied:
changed the location of the keys to `/usr/share/keyrings`.

Botspot commented on api:
```diff
@@ -351,6 +351,38 @@ app_to_pkgname() { #given an app-name, convert it to a unique, valid package-nam
+  echo "deb [signed-by=/tmp/pi-apps-apt-keys/$reponame-keyring.gpg] $repopath" | sudo tee /etc/apt/sources.list.d/${reponame}.list >/dev/null || error "add_external_repo: failed to create .list file"
+
+  # test if the repo works
+  sudo apt-get update -qq || error "add_external_repo: apt-get update failed. something is wrong with apt on your system."
```
I don't think it is good to have apt update within this function. `install_packages` will likely be run right after this function, and it already does an apt update anyway.

ryanfortner replied:
removed.

theofficialgman commented on api:
```diff
@@ -351,6 +351,43 @@ app_to_pkgname() { #given an app-name, convert it to a unique, valid package-nam
+  wget -q --spider "$pubkeyurl" || error "add_external_packages: pubkeyurl isn't a valid link"
+
+  # make pi-apps key storage dir if it doesn't exist
+  if [ ! -d /opt/pi-apps-apt-keys ]; then
+    mkdir -p /opt/pi-apps-apt-keys || error "add_external_repo: failed to create pi-apps key storage dir."
+  fi
```
oops... I had this review pending so you didn't see my comments from earlier today

my bad

theofficialgman commented on api:
```diff
@@ -351,6 +351,43 @@ app_to_pkgname() { #given an app-name, convert it to a unique, valid package-nam
+
+  # make pi-apps key storage dir if it doesn't exist
+  if [ ! -d /usr/share/keyrings ]; then
+    mkdir -p /usr/share/keyrings || error "add_external_repo: failed to create pi-apps key storage dir."
```
all files/folders in /usr are root owned, this would fail without sudo if it gets called

ryanfortner commented:
I just made a few changes. If everything looks good, I can go ahead and change the scripts for each of the apps that need the functions in this branch, so we can get it all done at once.

theofficialgman commented on api:
```diff
@@ -351,6 +351,55 @@ app_to_pkgname() { #given an app-name, convert it to a unique, valid package-nam
+
+  # download gpg key from specified url
+  rm -f /usr/share/keyrings/${reponame}-keyring.gpg
+  curl -fsSL "$pubkeyurl" | sudo gpg --dearmor -o /usr/share/keyrings${reponame}-keyring.gpg
```
missing `/`

theofficialgman commented on api:
```diff
@@ -351,6 +351,55 @@ app_to_pkgname() { #given an app-name, convert it to a unique, valid package-nam
+  fi
+
+  # create .list file
+  echo "deb [signed-by=/usr/share/keyrings/$reponame-keyring.gpg] $repopath" | sudo tee /etc/apt/sources.list.d/${reponame}.list >/dev/null
```
I think we still shouldn't specify a specific key. apt will search the whole directory for a valid key. in the chance that a repo updates its own key (in a different filename) the key that is hardcoded here may be invalid

theofficialgman commented on api:
```diff
@@ -351,6 +351,55 @@ app_to_pkgname() { #given an app-name, convert it to a unique, valid package-nam
+  fi
+}
+
+rm_external_repo() { # remove an external apt repo and its gpg key
+  reponame="$1"
+
+  rm -f /usr/share/keyrings/${reponame}-keyring.gpg || error "rm_external_repo: removal of ${reponame}-keyring.gpg failed"
+
+  sudo rm -f /etc/apt/sources.list.d/${reponame}.list || error "rm_external_repo: removal of ${reponame}.list failed"
+}
```
I still think this function should be made to be much more like "remove_repofile_if_unused" (thats an existing function)

we shouldn't always remove a repo because it may be in use in multiple apps.

ryanfortner replied:
makes sense. I will make some changes later tomorrow, it's getting late here.

ryanfortner commented:
Alright @theofficialgman I changed the `rm_external_repo` function to be more like the `remore_repofile_if_unused` function. 

Just in case the repo would need to be removed no matter what, I added a `force` option where the repo will be removed regardless of what packages are installed from it. (see the top comment)

theofficialgman commented:
this might be overkill, but spaces are not allowed in apt repo filename or keyname. this function doesn't do anything to prevent the script writer from making that mistake or document it.

at minimum some documents should be made to specify that repopath (the second input) should never contain spaces

ryanfortner commented:
I added a check for spaces and updated documentation.

Botspot commented on api:
```diff
@@ -351,6 +350,94 @@ app_to_pkgname() { #given an app-name, convert it to a unique, valid package-nam
+    sudo mkdir -p /usr/share/keyrings || error "add_external_repo: failed to create pi-apps key storage dir."
+  fi
+
+  # check if .list file already exists
+  [ -f /etc/apt/sources.list.d/${reponame}.list ] && error "add_external_repo: /etc/apt/sources.list.d/${reponame}.list already exists."
```
I feel like this error would be triggered if this function is somehow run twice, which is likely to happen.
Multiple apps may use the same repo, and I try to support the same app being installed multiple times without an uninstall in between.

Botspot replied:
Probably best to first remove it.

Botspot commented on api:
```diff
@@ -351,6 +350,94 @@ app_to_pkgname() { #given an app-name, convert it to a unique, valid package-nam
+  # check if .list file already exists
+  [ -f /etc/apt/sources.list.d/${reponame}.list ] && error "add_external_repo: /etc/apt/sources.list.d/${reponame}.list already exists."
+
+  # download gpg key from specified url
+  rm -f /usr/share/keyrings/${reponame}-keyring.gpg
+  curl -fsSL "$pubkeyurl" | sudo gpg --dearmor -o /usr/share/keyrings/${reponame}-keyring.gpg
```
Also, `sudo` is used to create the ${reponame}-keyring.gpg file, but no sudo is used to remove it. I haven't tested this, but it seems like rm would fail.

theofficialgman commented:
marking this as draft since I have been doing some testing and found that some repos need the signed-by to specify the key location, and some others don't need dearmor... this is going to be complicated to do universally because of this and it might be best to have a set of guidelines and something done on a script by script basis

theofficialgman commented:
Deb822 format (.sources) has been supported for many years now (and all modern distros are switching to it) and it is probably best that if we were to add functions for adding and removing repos that we also do this with the Deb822 format rather than the older .list format.

https://manpages.debian.org/bookworm/apt/sources.list.5.en.html#THE_DEB_AND_DEB-SRC_TYPES:_GENERAL_FORMAT

One nice thing about deb822 is it supported external public key files as well as embedding the public key directly within the .sources file.

theofficialgman commented:
I have revived this PR.
However the test case in the initial comment https://github.com/Botspot/pi-apps/pull/1902#issue-1251632420 will not work as it does not have a debian suite and component defined which is required according to the debian spec for third party repos which these new functions follow.

theofficialgman commented:
A new test case:
```bash
add_external_repo "https://adoptium.jfrog.io/artifactory/deb" "adoptium" "bionic" "main" "https://adoptium.jfrog.io/artifactory/api/security/keypair/default-gpg-key/public" || exit 1

rm_external_repo "adoptium" || exit 1 # this will remove the repo if nothing is used in it
rm_external_repo "adoptium" "force" || exit 1 # this will remove the repo even if software is installed from it
```

theofficialgman commented:
I have switched all applications that can be switched to use these generic functions.

The other applications (vivaldi and brave) cannot use these generic functions as they use their own custom non-standard key and repofile names and update them themselves as part of other debs.

theofficialgman commented:
A new test case (based on the above edits):
```bash
add_external_repo "adoptium" "https://adoptium.jfrog.io/artifactory/api/security/keypair/default-gpg-key/public" "https://adoptium.jfrog.io/artifactory/deb" "bionic" "main" || exit 1

rm_external_repo "adoptium" || exit 1 # this will remove the repo if nothing is used in it
rm_external_repo "adoptium" "force" || exit 1 # this will remove the repo even if software is installed from it
```

the original test case now becomes:
```bash
add_external_repo "box64" "https://ryanfortner.github.io/box64-debs/KEY.gpg" "https://ryanfortner.github.io/box64-debs/debian" "./" || exit 1

rm_external_repo "box64" || exit 1 # this will remove the repo if nothing is used in it
rm_external_repo "box64" "force" || exit 1 # this will remove the repo even if software is installed from it
```

theofficialgman commented:
@ryanfortner thanks for your original contribution! I will merge the current iteration of this PR. Sorry it took so long to get back around to this.

Between when this PR was opened and now pi-apps moved documentation from this repo to our website so I will rewrite that information there soon.
